_posts/2023-08-29-grpc-in-scala-with-fs2grpc-and-scalapb.md

```diff
@@ -0,0 +1,966 @@
+        note: "Samsung Galaxy Note 23",
+        a23: "Samsung Galaxy A23 5G ",
+        a54: "Samsung Galaxy A54",
+        a42: "Samsung A42 5G",
+        edgeNew: "Motorola Moto Edge+",
+        razr: "Motorola Moto Razr+",
+        think: "Motorola ThinkPhone",
+        edgeOld: "Motorola Moto Edge",
+        stylus: "Moto G Stylus 5G",
+        motog: "Moto G Power",
+        fiveiv: "Sony Xperia 5 IV",
+        onev: "Sony Xperia 1 V",
+        pro: "Sony Xperia Pro-1",
+        lfour: "Sony Xperia L4",
+        teniv: "Sony Xperia 10 IV",
+        tenv: "Sony Xperia 10 V",
+      };
+
+const forms = document.getElementsByTagName("form");
+
+const myButton = document.getElementById("submitBtn");
+const replyDiv = document.getElementById("reply");
+
+myButton.addEventListener("click", (event) => {
+  var orderObj = { orders: [] };
+  for (let i = 0; i < forms.length; i++) {
+    const formData = new FormData(forms[i]);
+    const jsonObject = Object.fromEntries(formData);
+    var jsonFormatted = {
+      items: [
+        {
+          name: names[jsonObject["flagship"]],
+          quantity: jsonObject["flagshipQty"],
+          amount: prices[jsonObject["flagship"]],
+        },
+        {
+          name: names[jsonObject["budget"]],
+          quantity: jsonObject["budgetQty"],
+          amount: prices[jsonObject["budget"]],
+        },
+      ],
+    };
+    orderObj.orders.push(jsonFormatted);
+  }
+
+  async function sendForm() {
+    const res = await fetch("http://localhost:8080/submit", {
+      method: "POST",
+      body: JSON.stringify(orderObj),
+    });
+    const resData = await res.json();
+    return resData;
+  }
+  return sendForm()
+    .then((arr) => arr.map((str) => {
+      var p = document.createElement('p')
+      p.textContent = str;
+      p.style.color = "darkorchid";
+      replyDiv.appendChild(p);
+    }))
+    .catch((err) => (replyDiv.innerHTML = err));
+});
+```
+## 7. The main program
+In this section, we bring everything together to run our program. First, we'll need a server to run `restService` that we created previously. Create `Main.scala` in the following path, `src/main/scala/Main.scala`, and add the following code.
+
+```scala
+import cats.effect.*
+import fs2.Stream
+import org.http4s.ember.server.EmberServerBuilder
+import com.comcast.ip4s.*
+import com.xonal.routes.MyRoutes.restService
+
+object Main extends IOApp {
+  def httpServerStream = Stream.eval(
+    EmberServerBuilder
+      .default[IO]
+      .withHost(host"0.0.0.0")
+      .withPort(port"8080")
+      .withHttpApp(restService.orNotFound)
+      .build
+      .use(_ => IO.never)
+  )
+}
+```
+We use `EmberServerBuilder` to create our server with port number 8080 and build it with our `restService`. 
+
+```scala
+import com.xonal.service.Server.grpcServer
+
+object Main extends IOApp {
+  ...
+  val grpcServerStream = Stream.eval(grpcServer)
+
+  def run(args: List[String]): IO[ExitCode] = grpcServerStream
```

Comment:
We can use the two resources in parallel by using `.parMapN` with IOs, rather than wrapping them into streams.
Not a hard opinion here though.

Reply:
Decided to use this